src/MongoDB.Driver.Core/Core/Bindings/CoreServerSessionPool.cs

```diff
@@ -62,31 +53,16 @@ public ICoreServerSession AcquireSession()
         /// <inheritdoc />
         public void ReleaseSession(ICoreServerSession session)
         {
-            lock (_lock)
+            if (IsAboutToExpire(session))
             {
-                var removeCount = 0;
-                for (var i = 0; i < _pool.Count; i++)
-                {
-                    var pooledSession = _pool[i];
-                    if (IsAboutToExpire(pooledSession))
-                    {
-                        pooledSession.Dispose();
-                        removeCount++;
-                    }
-                    else
-                    {
-                        break;
-                    }
-                }
-                _pool.RemoveRange(0, removeCount);
-
-                if (IsAboutToExpire(session))
-                {
-                    session.Dispose();
-                }
-                else
+                session.Dispose();
+            }
+            else
+            {
+                ICoreServerSession expiredSession;
+                if (_pool.ReleaseAndTryRemoveMatchingLeastRecentlyUsed(session, IsAboutToExpire, out expiredSession))
```

Comment:
Having a method that does two things:

1. Calls Release
2. Calls TryRemoveMatchingLeastRecentlyUsed

is a locking optimization. By doing both things while holding the lock we cut in half the number of times we have to acquire the lock, which reduces contention a lot.

Reply:
Note: this method used to be responsible for calling Dispose on as many expired sessions as possible. Now it's only responsible for calling Dispose on one at most. This distributes the work of disposing of the sessions among more threads.